### PR TITLE
refactor(core/ui): optimize brightness settings for Mercury

### LINKED ISSUE #3969: Brightness settings flow
**Design:** [Figma](https://www.figma.com/design/n43caWJFGJlT167AQcxruu/%E3%80%94-Prod-%E3%80%95Safe-5?node-id=6308-903&t=z3QeXzKofY9E1oQZ-4)

### PR DESCRIPTION
**Pull request by ibz**
Fixes #3969, #3962.

### CODE REVIEW

**Review by mmilata — COMMENTED** [review]
Some preliminary comments, so far looking good!

If you want to learn more about testing infrastructure feel free to write a [device test](https://docs.trezor.io/trezor-firmware/tests/index.html#device-tests) for this feature.

**Comment by Hannsek** [comment]
@ibz please post here some screenshots from the flow.

File: core/embed/rust/src/ui/model_mercury/layout.rs [hunk 1]
```diff
@@ -1026,14 +1018,104 @@ extern "C" fn new_select_word(n_args: usize, args: *const Obj, kwargs: *mut Map)
...
+            (SetBrightnessAction::Confirm, FlowMsg::Confirmed) => Decision::Return(FlowMsg::Confirmed),
+            (SetBrightnessAction::Confirm, FlowMsg::Info) => {
+                Decision::Goto(SetBrightnessAction::Menu, SwipeDirection::Left)
+            }
+            _ => Decision::Nothing,
+        }
+    }
+}
+
+static mut BRIGHTNESS: u16 = 0;
```
[the remark this answers is not in the document]
Comment (obrusvit) [reply core/embed/rust/src/ui/model_mercury/layout.rs:1065]: I also suggest indicating that `BRIGHTNESS` is in values of `[0-100] %`.

Either by a short commentary above it or changing the name to `BRIGTHNESS_PCT`

**Comment by ibz** [comment]
> @ibz please post here some screenshots from the flow.

image
image
image

File: core/embed/rust/src/ui/model_mercury/layout.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibz) [reply core/embed/rust/src/ui/model_mercury/layout.rs]: > I thing we can avoid the unsafe by using `core::sync::atomic::AtomicU16` instead.

Indeed! Thank you!

https://github.com/trezor/trezor-firmware/pull/3991/commits/9cfa654b31ce9528782e4506c003b5b799d3b7e0#diff-369a97745e9678828b91bb11d8de987cfac308ea29ed1702009e1f58d16fd45aR72

File: core/embed/rust/src/ui/model_mercury/layout.rs [hunk 2]
```diff
@@ -1026,14 +1018,104 @@ extern "C" fn new_select_word(n_args: usize, args: *const Obj, kwargs: *mut Map)
...
+            .with_footer(TR::instructions__tap_to_confirm.into(), None)
+            .with_menu_button()
+            .with_swipe(SwipeDirection::Down, SwipeSettings::default())
+            .with_swipe(SwipeDirection::Left, SwipeSettings::default())
+            .map(move |msg| match msg {
+                FrameMsg::Content(()) => {
+                    match storage::set_brightness(unsafe { BRIGHTNESS } as _) {
+                        Ok(_) => Some(FlowMsg::Confirmed),
+                        Err(_) => Some(FlowMsg::Cancelled),
+                    }
```
[the remark this answers is not in the document]
Comment (ibz) [reply core/embed/rust/src/ui/model_mercury/layout.rs]: https://github.com/trezor/trezor-firmware/pull/3991/commits/9cfa654b31ce9528782e4506c003b5b799d3b7e0#diff-369a97745e9678828b91bb11d8de987cfac308ea29ed1702009e1f58d16fd45aR122

File: core/embed/rust/src/ui/model_mercury/component/set_brightness.rs [hunk 3]
```diff
@@ -38,14 +36,7 @@ impl Component for SetBrightnessDialog {
...
                 display::backlight(value as _);
-                None
-            }
-            Some(NumberInputSliderDialogMsg::Cancelled) => Some(CancelConfirmMsg::Cancelled),
-            Some(NumberInputSliderDialogMsg::Confirmed) => {
-                match storage::set_brightness(self.0.value() as _) {
-                    Ok(_) => Some(CancelConfirmMsg::Confirmed),
-                    Err(_) => Some(CancelConfirmMsg::Cancelled), // TODO: handle error
-                }
+                Some(NumberInputSliderDialogMsg::Changed(value))
```
[the remark this answers is not in the document]
Comment (ibz) [reply core/embed/rust/src/ui/model_mercury/component/set_brightness.rs]: Indeed! https://github.com/trezor/trezor-firmware/pull/3991/commits/9cfa654b31ce9528782e4506c003b5b799d3b7e0#diff-f42b398908843fdcf7a5550c599833a44ccdad4b8bb64ec06860d0c6bb4fbfd1L1

File: core/embed/rust/src/ui/model_mercury/layout.rs [hunk 4]
```diff
@@ -1026,14 +1018,104 @@ extern "C" fn new_select_word(n_args: usize, args: *const Obj, kwargs: *mut Map)
     unsafe { util::try_with_args_and_kwargs(n_args, args, kwargs, block) }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq, ToPrimitive)]
+pub enum SetBrightnessAction {
```
[the remark this answers is not in the document]
Comment (ibz) [reply core/embed/rust/src/ui/model_mercury/layout.rs]: Much better! https://github.com/trezor/trezor-firmware/pull/3991/commits/9cfa654b31ce9528782e4506c003b5b799d3b7e0#diff-369a97745e9678828b91bb11d8de987cfac308ea29ed1702009e1f58d16fd45aR1

File: core/embed/rust/src/ui/model_mercury/layout.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibz) [reply core/embed/rust/src/ui/model_mercury/layout.rs]: > I also suggest indicating that `BRIGHTNESS` is in values of `[0-100] %`.
> 
> Either by a short commentary above it or changing the name to `BRIGTHNESS_PCT`

Actually this is `[0..255]`, not a percentage. It gets converted to percentage when displayed. But thanks to your comment I realized we should be just using u8 rather than u16 everywhere related to brightness. I'll make a separate PR on top of this.

File: core/translations/en.json [hunk 5]
```diff
@@ -81,7 +81,7 @@
     "bitcoin__unverified_external_inputs": "The transaction contains unverified external inputs.",
     "bitcoin__valid_signature": "The signature is valid.",
     "bitcoin__voting_rights": "Voting rights to:",
-    "brightness__title": "Set brightness",
+    "brightness__title": "Change display brightness",
```
Comment (mmilata) [inline core/translations/en.json:84]: @obrusvit do you know whether we need to do anything when an english string changes?

**Review by mmilata — COMMENTED** [review]
Code-wise this looks good, however the appearance is different from [figma](https://www.figma.com/design/n43caWJFGJlT167AQcxruu/%E3%80%94-Prod-%E3%80%95Safe-5?node-id=6308-903&t=z3QeXzKofY9E1oQZ-4). Perhaps it changed in the meanwhile? It has orange heading while the finished flows have green so maybe it's still not final?

**Comment by mmilata** [comment]
Don't know if you know but you can take screenshots by pressing `p` in the emulator, they're saved under `core/src/`.

**Comment by ibz** [comment]
New screenshots here @Hannsek :

emu00000000
emu00000001
emu00000002
emu00000003
emu00000004
emu00000005

**Comment by Hannsek** [comment]
Thanks, looks better now. But please make sure to have the right copy on right places and the right dimensions according to the [design](https://www.figma.com/design/n43caWJFGJlT167AQcxruu/%E3%80%94-Prod-%E3%80%95Safe-5?node-id=6308-903&t=fDsNN6rughEmbLfN-4).

File: core/embed/rust/src/ui/model_mercury/component/number_input_slider.rs [hunk 6]
```diff
@@ -72,17 +101,11 @@ impl Component for NumberInputSliderDialog {
...
-
-        shape::Text::new(self.text_area.center(), &str)
-            .with_font(theme::TEXT_NORMAL.text_font)
-            .with_fg(theme::TEXT_NORMAL.text_color)
-            .with_align(Alignment::Center)
-            .render(target);
+        if self.val == self.init_val || self.input.inner().touching {
+            self.init_footer.render(target);
+        } else {
+            self.modified_footer.render(target);
```
Comment (obrusvit) [inline core/embed/rust/src/ui/model_mercury/component/number_input_slider.rs:107]: Have you considered creating just one `footer` and using `self.footer.update_instruction(...)` and `self.footer.update_description(...)` instead of creating two `Footers` and switching their rendering?

File: core/translations/en.json (hunk 5 above)
Comment (obrusvit) [reply core/translations/en.json]: I'm not sure. In this case, it seems unnecessary. WDYT @Hannsek

Comment (Hannsek) [reply core/translations/en.json]: If it is not a big pain, please change it. 🙏🏻

Comment (ibz) [reply core/translations/en.json]: Will need new translations, for sure. But we are adding new strings anyway, so I think it's no big deal.

File: core/embed/rust/src/ui/model_mercury/component/number_input_slider.rs (hunk 6 above)
Comment (ibz) [reply core/embed/rust/src/ui/model_mercury/component/number_input_slider.rs]: > Have you considered creating just one `footer` and using `self.footer.update_instruction(...)` and `self.footer.update_description(...)` instead of creating two `Footers` and switching their rendering?

Much better, indeed!

https://github.com/trezor/trezor-firmware/pull/3991/commits/d2c19add05de1c5d68f7bc28c876dbdeac087553#diff-6e51ef35492aeba10734987923ac867ce64c62e70b23c8a78e47ee86c122c4b6R35

File: vendor/micropython
Comment (mmilata) [inline vendor/micropython:1]: Looks like this shouldn't be part of this PR?

File: core/embed/rust/src/ui/model_mercury/component/number_input_slider.rs [hunk 7]
```diff
@@ -135,20 +129,27 @@ impl NumberInputSlider {
...
     }
 
-    pub fn slider_eval(&mut self, pos: Point, ctx: &mut EventCtx) -> Option<u16> {
+    pub fn touch_eval(
+        &mut self,
+        pos: Point,
+        ctx: &mut EventCtx,
+        force_bubble_up: bool,
+    ) -> Option<u16> {
         if self.touch_area.contains(pos) {
```
Comment (mmilata) [inline core/embed/rust/src/ui/model_mercury/component/number_input_slider.rs:146]: I wonder if this condition should be replaced with `self.touching`. This would make it possible to change the value even when finger slides oudside rectangle, provided the TouchStart happened inside. Thinking of situation where you notice the footer changed and move your finger to read it, fat fingers, etc ...

**Review by mmilata — COMMENTED** [review]
Almost ready to merge :+1:

File: core/embed/rust/src/ui/model_mercury/component/number_input_slider.rs (hunk 7 above)
Comment (ibz) [reply core/embed/rust/src/ui/model_mercury/component/number_input_slider.rs]: Indeed, it feels better ... https://github.com/trezor/trezor-firmware/pull/3991/commits/cf5be6ff82a54b5abf521a1f9877cc5bc9069663#diff-6e51ef35492aeba10734987923ac867ce64c62e70b23c8a78e47ee86c122c4b6R146

**Review by mmilata — APPROVED** [review]
Please squash your fixups and then merge if CI passes.

**Comment by ibz** [comment]
Monero test failed. Ran it separately and it passed. Merging...
